### PR TITLE
ISSUE-2640: BP-43 Integrate spotbugs plugin with gradle

### PR DESCRIPTION
**Pull request by pkumar-singh**
### Motivation

**Migrate bookkeeper to gradle spotbug plugin.**
spotbugs is gradle as well as maven plugin which checks for obvious bugs in source code.
**How to run** 
 `./gradlew spotbugsMain `

### Changes

- Integrate gradle build with spotbugs plugin
- Second commit is for fixing issues reported by spotbugs plugin 

### Test
Just to validate it's working. 
I introduced following bug.
 ```
public void test() {
+
+        String[] field = {"a", "b", "c", "s", "e"};
+
+        //concatenates strings using + in a loop
+        String ip = "127.0.0.1";
+        String s = "";
+        for (int i = 0; i < field.length; ++i) {
+            s = s + field[i];
+        }
+
+        System.out.println(ip);
+
+    }
```

spotbugs plugin reported following violation.
```
<BugInstance type="SBSC_USE_STRINGBUFFER_CONCATENATION" priority="2" rank="18" abbrev="SBSC" category="PERFORMANCE" instanceHash="2aa8e4ebcc4cd86190fe3e7bd8a4913c" instanceOccurrenceNum="0" instanceOccurrenceMax="0">
    <ShortMessage>Method concatenates strings using + in a loop</ShortMessage>
    <LongMessage>org.apache.bookkeeper.meta.LongZkLedgerIdGenerator.test() concatenates strings using + in a loop</LongMessage>
    <Class classname="org.apache.bookkeeper.meta.LongZkLedgerIdGenerator" primary="true">
      <SourceLine classname="org.apache.bookkeeper.meta.LongZkLedgerIdGenerator" start="58" end="350" sourcefile="LongZkLedgerIdGenerator.java" sourcepath="org/apache/bookkeeper/meta/LongZkLedgerIdGenerator.java" relSourcepath="java/org/apache/bookkeeper/meta/LongZkLedgerIdGenerator.java">
        <Message>At LongZkLedgerIdGenerator.java:[lines 58-350]</Message>
      </SourceLine>
      <Message>In class org.apache.bookkeeper.meta.LongZkLedgerIdGenerator</Message>
    </Class>
```
Master Issue: #2640

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -19,4 +19,5 @@
 #
 #
 
-protobufPluginVersion=0.8.15
\ No newline at end of file
+protobufPluginVersion=0.8.15
+spotbugsPlugin=4.7.0
```
Comment (eolivelli) [inline gradle.properties:23]: is this the same version that we have in Maven pom.xml ?
why the new spotbugs found issues that were not already tracked?

can you please update the pom.xml as well ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
overall +1
I left one comment about the version of spotbugs gradle vs Maven

File: gradle.properties (hunk 1 above)
Comment (pkumar-singh) [reply gradle.properties:23]: Yeah. That's version of spotbugs gradle plugin itself. This is version for the tool version such as engine/algorithm to find bugs which is  toolVersion = '3.1.8'.    <spotbugs-annotations.version>3.1.8</spotbugs-annotations.version>

Comment (hsaputra) [reply gradle.properties]: Hi @eolivelli - could you help review this again? Thanks

**Review by hsaputra — APPROVED** [review]
LGTM +1

**Review by eolivelli — APPROVED** [review]
+1

**Comment by merlimat** [comment]
@pkumar-singh You might need to rebase on master to get fixes for CI

**Comment by hsaputra** [comment]
The failed tests seems coming from timeouts and repeating fails in other PRs.
Since the change is related gradle changes, will merge this one.

**Pull request merged**